### PR TITLE
Add RPC End Point Mapper

### PR DESCRIPTION
**Pull request by olbjo**
Based on contributions in https://github.com/rtlabs-com/p-net/pull/151
by atomic-3dpenguin

-------------------------
Epm check in user scenario 10 passes
Still problems with IP_UPD_RPC_I&M_EPM test case

### CODE REVIEW

File: src/device/pf_cmrpc_helpers.c [hunk 1]
```diff
@@ -53,3 +98,309 @@ void pf_generate_uuid (
...
+ * @param p_lookup_req     In:    Lookup request.
+ * @param p_lookup_rsp     Out:   Lookup response.
+ * @param p_read_status    Out:   Read pnio status
+ * @param res_size         In:    The size of the output buffer.
+ * @param p_res            Out:   The output buffer.
+ * @param p_pos            InOut: Position in the output buffer.
+ * @return  0  if operation succeeded.
+ *          -1 if an error occurred.
+ */
+static int pf_cmrdr_add_epmv4_entry (
```
Comment (pyhys) [inline src/device/pf_cmrpc_helpers.c:191]: The documentation does not seem to match. Please make input pointer arguments const.

File: src/device/pf_cmrpc_helpers.c [hunk 2]
```diff
@@ -53,3 +98,309 @@ void pf_generate_uuid (
...
+ * @param p_lookup_req     In:    Lookup request.
+ * @param p_lookup_rsp     Out:   Lookup response.
+ * @param p_read_status    Out:   Read pnio status
+ * @return  0  if operation succeeded.
+ *          -1 if an error occurred.
+ */
+static int pf_cmrdr_add_pnio_entry (
+   pnet_t * net,
+   pf_rpc_header_t * p_rpc_req,
+   pf_rpc_lookup_req_t * p_lookup_req,
```
Comment (pyhys) [inline src/device/pf_cmrpc_helpers.c:235]: These two should be const

File: src/device/pf_cmrpc_helpers.c [hunk 3]
```diff
@@ -53,3 +98,309 @@ void pf_generate_uuid (
...
+
+   p_lookup_rsp->return_code = PF_RPC_OK;
+   return 0;
+}
+
+/**
+ * Handle epm read all inquiry request
+ * PN-AL-protocol (Mar20) Section 4.10.3
+ *
+ * PN-AL-protocol (Mar20) Figure 3
```
Comment (pyhys) [inline src/device/pf_cmrpc_helpers.c:286]: Figure 3?

File: src/device/pf_cmrpc_helpers.c [hunk 4]
```diff
@@ -53,3 +98,309 @@ void pf_generate_uuid (
...
+ * @param p_lookup_req     In:    Lookup request.
+ * @param p_lookup_rsp     Out:   Lookup response.
+ * @param p_read_status    Out:   Read pnio status
+ * @return  0  if operation succeeded.
+ *          -1 if an error occurred.
+ */
+static int pf_cmrdr_inquiry_read_all_reg_ind (
+   pnet_t * net,
+   pf_rpc_header_t * p_rpc_req,
+   pf_rpc_lookup_req_t * p_lookup_req,
```
Comment (pyhys) [inline src/device/pf_cmrpc_helpers.c:301]: const

File: src/device/pf_cmrpc_helpers.c [hunk 5]
```diff
@@ -53,3 +98,309 @@ void pf_generate_uuid (
...
+ * @param res_size         In:    The size of the output buffer.
+ * @param p_res            Out:   The output buffer.
+ * @param p_pos            InOut: Position in the output buffer.
+ * @return  0  if operation succeeded.
+ *          -1 if an error occurred.
+ */
+int pf_cmrpc_lookup_request (
+   pnet_t * net,
+   pf_rpc_header_t * p_rpc_req,
+   pf_rpc_lookup_req_t * p_lookup_req,
```
Comment (pyhys) [inline src/device/pf_cmrpc_helpers.c:367]: two const

File: src/pf_types.h [hunk 6]
```diff
@@ -216,6 +216,145 @@ typedef enum pf_write_req_ext_error_type_values
    PF_WRT_ERROR_NO_PEER_DETECTED = 0x8005
 } pf_write_req_ext_error_type_t;
 
+/*
+ * RPC and EPM implementation
+ * PN-AL-protocol (Mar20) Section 4.10.3
+ */
+
+#define RPC_TOWER_REFERENTID  3
```
Comment (pyhys) [inline src/pf_types.h:224]: PF_RPC_TOWER_REFERENTID

File: src/device/pf_block_writer.c [hunk 7]
```diff
@@ -135,6 +135,25 @@ static void pf_put_uuid (
    pf_put_mem (p_uuid->data4, sizeof (p_uuid->data4), res_len, p_bytes, p_pos);
 }
 
+static void pf_put_rpc_uuid (
```
Comment (pyhys) [inline src/device/pf_block_writer.c:138]: Add documentation

File: src/device/pf_block_writer.c [hunk 8]
```diff
@@ -135,6 +135,25 @@ static void pf_put_uuid (
...
+   bool is_big_endian,
+   pf_rpc_uuid_type_t * p_uuid,
```
Comment (pyhys) [inline src/device/pf_block_writer.c:140]: const

File: src/device/pf_block_writer.c [hunk 9]
```diff
@@ -4109,3 +4128,442 @@ void pf_put_pdport_data_adj (
...
+ * Put rpc epm floor 1 configuration into a buffer
+ * @param is_big_endian    In:   Endianness of the destination buffer.
+ * @param p_floor          In:   Floor configuration.
+ * @param res_len          In:   Size of destination buffer.
+ * @param p_bytes          Out:  Destination buffer.
+ * @param p_pos            InOut:Position in destination buffer.
+ */
+static void pf_put_rpc_floor_1_uuid (
+   bool is_big_endian,
+   pf_floor_uuid_t * p_floor,
```
Comment (pyhys) [inline src/device/pf_block_writer.c:4160]: const

File: src/device/pf_block_writer.c [hunk 10]
```diff
@@ -4109,3 +4128,442 @@ void pf_put_pdport_data_adj (
...
+ * Put rpc epm floor 2 configuration into a buffer
+ * @param is_big_endian    In:   Endianness of the destination buffer.
+ * @param p_floor          In:   Floor configuration.
+ * @param res_len          In:   Size of destination buffer.
+ * @param p_bytes          Out:  Destination buffer.
+ * @param p_pos            InOut:Position in destination buffer.
+ */
+static void pf_put_rpc_floor_2_uuid (
+   bool is_big_endian,
+   pf_floor_uuid_t * p_floor,
```
Comment (pyhys) [inline src/device/pf_block_writer.c:4207]: const

File: src/device/pf_block_writer.c [hunk 11]
```diff
@@ -4109,3 +4128,442 @@ void pf_put_pdport_data_adj (
...
+ * Put rpc epm floor 3 configuration into a buffer
+ * @param is_big_endian    In:   Endianness of the destination buffer.
+ * @param p_floor          In:   Floor configuration.
+ * @param res_len          In:   Size of destination buffer.
+ * @param p_bytes          Out:  Destination buffer.
+ * @param p_pos            InOut:Position in destination buffer.
+ */
+static void pf_put_rpc_floor_3_rpc (
+   bool is_big_endian,
+   pf_floor_3_t * p_floor,
```
Comment (pyhys) [inline src/device/pf_block_writer.c:4225]: const

File: src/device/pf_block_writer.c [hunk 12]
```diff
@@ -4109,3 +4128,442 @@ void pf_put_pdport_data_adj (
...
+ * Put rpc epm floor 4 configuration into a buffer
+ * @param is_big_endian    In:   Endianness of the destination buffer.
+ * @param p_floor          In:   Floor configuration.
+ * @param res_len          In:   Size of destination buffer.
+ * @param p_bytes          Out:  Destination buffer.
+ * @param p_pos            InOut:Position in destination buffer.
+ */
+static void pf_put_rpc_floor_4_udp (
+   bool is_big_endian,
+   pf_floor_4_t * p_floor,
```
Comment (pyhys) [inline src/device/pf_block_writer.c:4264]: const

File: src/device/pf_block_writer.c [hunk 13]
```diff
@@ -4109,3 +4128,442 @@ void pf_put_pdport_data_adj (
...
+ * Put rpc epm floor 5 configuration into a buffer
+ * @param is_big_endian    In:   Endianness of the destination buffer.
+ * @param p_floor          In:   Floor configuration.
+ * @param res_len          In:   Size of destination buffer.
+ * @param p_bytes          Out:  Destination buffer.
+ * @param p_pos            InOut:Position in destination buffer.
+ */
+static void pf_put_rpc_floor_5_ip (
+   bool is_big_endian,
+   pf_floor_5_t * p_floor,
```
Comment (pyhys) [inline src/device/pf_block_writer.c:4298]: const

File: src/device/pf_block_writer.c [hunk 14]
```diff
@@ -4109,3 +4128,442 @@ void pf_put_pdport_data_adj (
...
+ * Blank
+ * HWRevision        VisibleString[5]     4.10.3.3.3
+ * Blank
+ * SWRevisionPrefix  VisibleString[1]     4.10.3.3.4
+ * SWRevision        VisibleString[9]     4.10.3.3.5
+ * EndTerm           '\0'
+ *
+ * Todo refactor after https://github.com/rtlabs-com/p-net/pull/211 is added
+ *
+ * @param is_big_endian    In:   Endianness of the destination buffer.
```
Comment (pyhys) [inline src/device/pf_block_writer.c:4340]: Remove line

File: src/device/pf_block_writer.c [hunk 15]
```diff
@@ -4109,3 +4128,442 @@ void pf_put_pdport_data_adj (
...
+ * @param p_im_0_data      In:   I&M data used to generate annotation .
+ * @param res_len          In:   Size of destination buffer.
+ * @param p_bytes          Out:  Destination buffer.
+ * @param p_pos            InOut:Position in destination buffer.
+ */
+static void pf_put_rpc_epm_annotation (
+   pnet_im_0_t * p_im_0_data,
```
Comment (pyhys) [inline src/device/pf_block_writer.c:4347]: const

File: src/device/pf_block_writer.c [hunk 16]
```diff
@@ -4109,3 +4128,442 @@ void pf_put_pdport_data_adj (
...
+ * Put rpc epm protocol tower into a buffer
+ * @param is_big_endian    In:   Endianness of the destination buffer.
+ * @param p_tower          In:   Protocol tower configuration.
+ * @param res_len          In:   Size of destination buffer.
+ * @param p_bytes          Out:  Destination buffer.
+ * @param p_pos            InOut:Position in destination buffer.
+ */
+static void pf_put_tower_entry (
+   bool is_big_endian,
+   pf_rpc_tower_t * p_tower,
```
Comment (pyhys) [inline src/device/pf_block_writer.c:4385]: const

File: src/device/pf_block_writer.c [hunk 17]
```diff
@@ -4109,3 +4128,442 @@ void pf_put_pdport_data_adj (
...
+ * Put rpc handle into a buffer
+ * @param is_big_endian    In:   Endianness of the destination buffer.
+ * @param p_handle         In:   Rpc handle
+ * @param res_len          In:   Size of destination buffer.
+ * @param p_bytes          Out:  Destination buffer.
+ * @param p_pos            InOut:Position in destination buffer.
+ */
+static void pf_put_rpc_handle (
+   bool is_big_endian,
+   pf_rpc_handle_t * p_handle,
```
Comment (pyhys) [inline src/device/pf_block_writer.c:4464]: const

File: src/device/pf_block_writer.c [hunk 18]
```diff
@@ -4109,3 +4128,442 @@ void pf_put_pdport_data_adj (
...
+ * @param is_big_endian    In:   Endianness of the destination buffer.
+ * @param p_entry          In:   Rpc epm entry including protocol
+ *                               tower configuration.
+ * @param res_len          In:   Size of destination buffer.
+ * @param p_bytes          Out:  Destination buffer.
+ * @param p_pos            InOut:Position in destination buffer.
+ */
+static void pf_put_rpc_epm_entry (
+   bool is_big_endian,
+   pf_rpc_entry_t * p_entry,
```
Comment (pyhys) [inline src/device/pf_block_writer.c:4496]: const

**Review by pyhys — COMMENTED** [review]
Excellent! Some trivial things should be changed. 
I have repeated the same review comment on multiple places, so it should be easy for you to modify it.

File: src/device/pf_cmrpc.c [hunk 19]
```diff
@@ -2707,6 +2719,63 @@ static int pf_cmrpc_rm_read_ind (
...
+ * @param res_size         In:    The size of the response buffer.
+ * @param p_res            Out:   The response buffer.
+ * @param p_res_pos        InOut: Position within the response buffer.
+ * @return  0  if operation succeeded.
+ *          -1 if an error occurred.
+ */
+static int pf_cmrpc_lookup_ind (
+   pnet_t * net,
+   pf_session_info_t * p_sess,
+   pf_rpc_header_t * p_rpc_req,
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/device/pf_cmrpc.c:2740]: done

File: src/device/pf_block_writer.h [hunk 20]
```diff
@@ -762,6 +762,24 @@ void pf_put_pd_real_data (
...
+ * @param p_lookup_rsp     In:    Epm lookup response including entry and
+ *                                protocol tower configuration.
+ * @param res_len          In:    Size of destination buffer.
+ * @param p_bytes          Out:   Destination buffer.
+ * @param p_pos            InOut: Position in destination buffer.
+ */
+void pf_put_lookup_response_data (
+   pnet_t * net,
+   bool is_big_endian,
+   pf_rpc_lookup_rsp_t * p_lookup_rsp,
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/device/pf_block_writer.h:778]: done

File: src/device/pf_block_reader.h [hunk 21]
```diff
@@ -236,6 +236,17 @@ void pf_get_read_request (
    uint16_t * p_pos,
    pf_iod_read_request_t * p_req);
 
+/**
+ * Extract a lookup request block from a buffer.
+ * @param p_info           In:   The parser information.
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/device/pf_block_reader.h:241]: done

File: src/device/pf_block_writer.c (hunk 14 above)
Comment (olbjo) [reply src/device/pf_block_writer.c:4340]: done

File: src/device/pf_block_writer.c (hunk 15 above)
Comment (olbjo) [reply src/device/pf_block_writer.c:4347]: done

File: src/device/pf_block_writer.c (hunk 17 above)
Comment (olbjo) [reply src/device/pf_block_writer.c:4464]: done

File: src/device/pf_block_writer.c (hunk 18 above)
Comment (olbjo) [reply src/device/pf_block_writer.c:4496]: done

**Comment by olbjo** [comment]
Fixed comments,
Moved epm implementation from pf_cmrpc_helpers to new separate files after discussion

File: src/device/pf_cmrpc_epm.h [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+ * @param res_size         In:    The size of the output buffer.
+ * @param p_res            Out:   The output buffer.
+ * @param p_pos            InOut: Position in the output buffer.
+ * @return  0  if operation succeeded.
+ *          -1 if an error occurred.
+ */
+int pf_cmrpc_lookup_request (
+   pnet_t * net,
+   const pf_rpc_header_t * p_rpc_req,
+   pf_rpc_lookup_req_t * p_lkup_request,
```
Comment (pyhys) [inline src/device/pf_cmrpc_epm.h:38]: const  
Name should be p_lookup_req (or change in docstring)

File: src/device/pf_block_writer.c (hunk 7 above)
Comment (pyhys) [reply src/device/pf_block_writer.c:138]: Add:
```
/**
 * Put rpc UUID into a buffer
 * @param is_big_endian    In:   Endianness of the destination buffer.
 * @param p_uuid           In:   The UUID to insert.
 * @param res_len          In:   Size of destination buffer.
 * @param p_bytes          Out:  Destination buffer.
 * @param p_pos            InOut:Position in destination buffer.
 */
```

File: src/device/pf_block_writer.c [hunk 23]
```diff
@@ -4109,3 +4128,441 @@ void pf_put_pdport_data_adj (
...
+   uint8_t * p_bytes,
+   uint16_t * p_pos)
+{
+   pf_put_uint32 (is_big_endian, p_entry->max_count, res_len, p_bytes, p_pos);
+
+   pf_put_uint32 (is_big_endian, p_entry->offset, res_len, p_bytes, p_pos);
+
+   pf_put_uint32 (is_big_endian, p_entry->actual_count, res_len, p_bytes, p_pos);
+
+   /*Add entry if there if not empty */
```
Comment (pyhys) [inline src/device/pf_block_writer.c:4506]: `/* Add entry if not empty */`

File: src/device/pf_block_writer.c [hunk 24]
```diff
@@ -4109,3 +4128,441 @@ void pf_put_pdport_data_adj (
...
+      is_big_endian,
+      &p_lookup_rsp->rpc_entry,
+      res_len,
+      p_bytes,
+      p_pos);
+
+   /*Add padding if needed*/
+   if (*p_pos % 2 > 0)
+   {
+      pf_put_byte (pad8, res_len, p_bytes, p_pos);
```
Comment (pyhys) [inline src/device/pf_block_writer.c:4559]: Maybe just `pf_put_byte (0, ...`  
I think that is more obvious to the reader.

File: src/device/pf_cmrpc_epm.c [hunk 25]
```diff
@@ -0,0 +1,373 @@
...
+}
+
+/**
+ * Add EPMv4 endpoint to lookup response
+ * @param net              InOut: The p-net stack instance
+ * @param p_rpc_req        In:    Rpc header.
+ * @param p_lookup_req     In:    Lookup request.
+ * @param p_lookup_rsp     Out:   Lookup response.
+ * @param p_read_status    Out:   Read pnio status
+ * @param res_size         In:    The size of the output buffer.
```
Comment (pyhys) [inline src/device/pf_cmrpc_epm.c:152]: Remove three lines.
The names `p_lookup_req` and `p_lookup_rsp` are different than in the code.

File: src/device/pf_cmrpc_epm.c [hunk 26]
```diff
@@ -0,0 +1,373 @@
...
+ * @param p_res            Out:   The output buffer.
+ * @param p_pos            InOut: Position in the output buffer.
+ * @return  0  if operation succeeded.
+ *          -1 if an error occurred.
+ */
+static int pf_cmrdr_add_epmv4_entry (
+   pnet_t * net,
+   const pf_rpc_header_t * p_rpc_req,
+   pf_rpc_lookup_req_t * p_lkup_request,
```
Comment (pyhys) [inline src/device/pf_cmrpc_epm.c:161]: const

File: src/device/pf_cmrpc_epm.c [hunk 27]
```diff
@@ -0,0 +1,373 @@
...
+ * @param p_lookup_req     In:    Lookup request.
+ * @param p_lookup_rsp     Out:   Lookup response.
+ * @param p_read_status    Out:   Read pnio status
+ * @return  0  if operation succeeded.
+ *          -1 if an error occurred.
+ */
+static int pf_cmrdr_add_pnio_entry (
+   pnet_t * net,
+   const pf_rpc_header_t * p_rpc_req,
+   pf_rpc_lookup_req_t * p_lookup_req,
```
Comment (pyhys) [inline src/device/pf_cmrpc_epm.c:202]: const

File: src/device/pf_cmrpc_epm.c [hunk 28]
```diff
@@ -0,0 +1,373 @@
...
+ * @param p_lookup_req     In:    Lookup request.
+ * @param p_lookup_rsp     Out:   Lookup response.
+ * @param p_read_status    Out:   Read pnio status
+ * @return  0  if operation succeeded.
+ *          -1 if an error occurred.
+ */
+static int pf_cmrdr_inquiry_read_all_reg_ind (
+   pnet_t * net,
+   const pf_rpc_header_t * p_rpc_req,
+   pf_rpc_lookup_req_t * p_lookup_req,
```
Comment (pyhys) [inline src/device/pf_cmrpc_epm.c:268]: const

**Review by pyhys — COMMENTED** [review]
I like the new file!
Just a few remaining issues.

File: src/device/pf_cmrpc_epm.h (hunk 22 above)
Comment (olbjo) [reply src/device/pf_cmrpc_epm.h:38]: done

File: src/device/pf_block_writer.c (hunk 23 above)
Comment (olbjo) [reply src/device/pf_block_writer.c:4506]: done

File: src/device/pf_block_writer.c (hunk 24 above)
Comment (olbjo) [reply src/device/pf_block_writer.c:4559]: done

File: src/device/pf_cmrpc_epm.c (hunk 25 above)
Comment (olbjo) [reply src/device/pf_cmrpc_epm.c:152]: done

File: src/device/pf_cmrpc_epm.c (hunk 26 above)
Comment (olbjo) [reply src/device/pf_cmrpc_epm.c:161]: done

File: src/device/pf_cmrpc_epm.c (hunk 27 above)
Comment (olbjo) [reply src/device/pf_cmrpc_epm.c:202]: done

File: src/device/pf_cmrpc_epm.c (hunk 28 above)
Comment (olbjo) [reply src/device/pf_cmrpc_epm.c:268]: done

File: src/device/pf_block_writer.c (hunk 23 above)
Comment (olbjo) [reply src/device/pf_block_writer.c:4506]: also added docstrings for pf_put_rpc_uuid but can't figure out how to respond to it

File: src/device/pf_block_writer.c (hunk 16 above)
Comment (olbjo) [reply src/device/pf_block_writer.c]: done

File: src/device/pf_block_reader.c [hunk 29]
```diff
@@ -157,6 +157,31 @@ static void pf_get_uuid (
    pf_get_mem (p_info, p_pos, sizeof (p_dest->data4), p_dest->data4);
 }
 
+/**
+ * @internal
+ * Extract a rpc epm handle from a buffer.
+ * @param p_info           In:   The parser state.
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/device/pf_block_reader.c:163]: Done

File: src/device/pf_block_writer.c (hunk 7 above)
Comment (pyhys) [reply src/device/pf_block_writer.c:149]: Ok now

File: src/device/pf_block_writer.c [hunk 30]
```diff
@@ -4109,3 +4139,443 @@ void pf_put_pdport_data_adj (
...
+ * Blank
+ * HWRevision        VisibleString[5]     4.10.3.3.3
+ * Blank
+ * SWRevisionPrefix  VisibleString[1]     4.10.3.3.4
+ * SWRevision        VisibleString[9]     4.10.3.3.5
+ * EndTerm           '\0'
+ *
+ * Todo refactor after https://github.com/rtlabs-com/p-net/pull/211 is added
+ *
+ * @param p_im_0_data      In:   I&M data used to generate annotation .
```
Comment (pyhys) [inline src/device/pf_block_writer.c:4351]: * @param  p_cfg   In: Configuration

File: src/device/pf_cmrpc_epm.c [hunk 31]
```diff
@@ -0,0 +1,370 @@
...
+   p_handle->handle_uuid.node[3] = mac_address.addr[3];
+   p_handle->handle_uuid.node[4] = mac_address.addr[4];
+   p_handle->handle_uuid.node[5] = mac_address.addr[5];
+}
+
+/**
+ * Initialize protocol tower configuration including floors
+ * @param p_tower    Out: Protocol tower configuration
+ * @param epm_type   In: End point type, EPM or PNIO
+ * @param udp_port   In: Upd protocol port
```
Comment (pyhys) [inline src/device/pf_cmrpc_epm.c:99]: UDP protocol port

File: src/device/pf_cmrpc_epm.c [hunk 32]
```diff
@@ -0,0 +1,370 @@
...
+
+   p_lookup_rsp->return_code = PF_RPC_OK;
+   return 0;
+}
+
+/**
+ * Handle epm read all inquiry request
+ * PN-AL-protocol (Mar20) Section 4.10.3
+ *
+ * PN-AL-protocol (Mar20) Figure 3
```
Comment (pyhys) [inline src/device/pf_cmrpc_epm.c:250]: Is this correct figure number?

File: src/device/pf_cmrpc_epm.c [hunk 33]
```diff
@@ -0,0 +1,370 @@
...
+ * @param res_size         In:    The size of the output buffer.
+ * @param p_res            Out:   The output buffer.
+ * @param p_pos            InOut: Position in the output buffer.
+ * @return  0  if operation succeeded.
+ *          -1 if an error occurred.
+ */
+int pf_cmrpc_lookup_request (
+   pnet_t * net,
+   const pf_rpc_header_t * p_rpc_req,
+   pf_rpc_lookup_req_t * p_lookup_req,
```
Comment (pyhys) [inline src/device/pf_cmrpc_epm.c:331]: const

File: src/pf_types.h (hunk 6 above)
Comment (pyhys) [reply src/pf_types.h:224]: Update to PF_RPC_TOWER_REFERENTID

File: src/device/pf_cmrpc_epm.c (hunk 32 above)
Comment (olbjo) [reply src/device/pf_cmrpc_epm.c:250]: Changed to PN_AL-services

File: src/device/pf_block_writer.c (hunk 30 above)
Comment (olbjo) [reply src/device/pf_block_writer.c:4351]: done

File: src/device/pf_cmrpc_epm.c (hunk 33 above)
Comment (olbjo) [reply src/device/pf_cmrpc_epm.c:331]: done

File: src/device/pf_cmrpc_epm.c (hunk 32 above)
Comment (olbjo) [reply src/device/pf_cmrpc_epm.c:250]: PN-AL-services...

**Review by pyhys — APPROVED** [review]

**Pull request merged**